kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs

```diff
@@ -731,22 +731,25 @@ impl<'tcx> GotocCtx<'tcx> {
                 ref dst,
                 ref count,
             }) => {
-                let src = self.codegen_operand(src).cast_to(Type::void_pointer());
-                let dst = self.codegen_operand(dst);
-                let count = self.codegen_operand(count);
-                let sz = dst.typ().base_type().unwrap().sizeof(&self.symbol_table);
-                let sz = Expr::int_constant(sz, Type::size_t());
-                let n = sz.mul(count);
-                let dst = dst.cast_to(Type::void_pointer());
-                let e = BuiltinFn::Memcpy.call(vec![dst, src, n.clone()], location.clone());
-
-                // The C implementation of memcpy does not allow an invalid pointer for
-                // the src/dst, but the LLVM implementation specifies that a copy with
-                // length zero is a no-op. This comes up specifically when handling
-                // the empty string; CBMC will fail on passing a reference to empty
-                // string unless we codegen this zero check.
-                // https://llvm.org/docs/LangRef.html#llvm-memcpy-intrinsic
-                Stmt::if_then_else(n.is_zero().not(), e.as_stmt(location.clone()), None, location)
+                // It's surprising that calls to the intrinsic
+                // `copy_nonoverlapping` are not handled in `codegen_intrinsic`.
+                // We pack the operands and their types, and then call
+                // `codegen_copy_intrinsic` as if it was an intrinsic.
+                let fargs = vec![
+                    self.codegen_operand(src),
+                    self.codegen_operand(dst),
+                    self.codegen_operand(count),
+                ];
+                let farg_types =
+                    &[self.operand_ty(src), self.operand_ty(dst), self.operand_ty(count)];
+                self.codegen_copy_intrinsic(
```

Comment:
It is a bit odd to have this code calling a codegen intrinsic. Any chance you can put the common logic in a a separate function?

Reply:
Sorry, I don't know what you mean. The common logic is already in `codegen_copy_intrinsic`, I'm not sure if you want to move it somewhere else or split it into several functions.

Reply:
Maybe just rename it to not have the word "intrinsic"?